enhancements/testing/build_root_golang_version_from_repo.md

```diff
@@ -0,0 +1,130 @@
+status: provisional
+---
+
+
+# Build root image from repository
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Today, the image in which binaries are built by ci-operator (`build_root_image`) can be
+configured by either specifying an imagestream or by specifying a Dockerfile in the
+repository to build this image from. In the latler case, the Dockerfile is in the basebranch
+of PRs. This means its not possible to do a PR that atomically changes both contents of the
+repository and the `build_root_image`. This proposal proposes a way to enable changing both
+code and the `build_root_image` atomically with one PR.
+
+
+## Motivation
+
+### Goals
+
+* Allow changing both repository contents and `build_root` in a single PR
+
+### Non-Goals
+
+* Change anything about the concept of the `build_root` or how we build images
+  for promotion
+
+## Background
+
+The ci-operator config has a `build_root` section that can either include an imagestreamtag
+or a Dockerfile. If it is an Dockerfile, a build of that Dockerfile from the base branch of the
+PR will be created. In both cases, the resulting image will be tagged into the `pipeline:root`
+imagestreamtag.
+
+The `pipeline:root` imagestreamtag in turn is used to:
+* Build binaries via the `binary_build_command` and tag them into the `pipeline:bin` imagestreamtag
+* Build test binaries via the `test_binary_build_commands` and tag them into the `pipeline:test-bin` imagestreamtag
+* Build rpms via the `rpm_build_commands` and tag them into the `pipeline:rpms` imagestreamtag
+
+All these tags in turn can then be refenced in tests.
+
+Additionally, if the project promotes, there is one Dockerfile per promoted image. It usually duplicates
+the `build_root` and `binary_build_commands` but copies the resulting binaries in an `ocp/$version:base`
+image.
+
+## Proposal
+
+It is proposed to extend the existing `build_root` section in the ci-operator config
+with a new `from_repository` boolean that will indicate of the image should be inferred
+from the repository:
+```
+build_root:
+	from_repository: true
+```
+
+This setting will be mutually exclusive with all other options in the `build_root` section.
+If set, the `ci-operator-prowgen` cli will be changed to clone the repository into the ci
+operator pod. This in turn allows the `ci-operator` to read the image to be used from the
+repository.
+
+The `build_root_image` must not be a raw docker pullspec but an imagestreamtag, because
+we tag it into the `pipeline:root` imagestreamtag and have found that doing an import for
+every job run causes performance issues in the underlying cluster.
+
+TODO: Agree on which of the two options below is better.
```

Comment:
Is the build_root_image imagestreamtag the same string as what we have in Dockerfiles? Like `registry.svc.ci.openshift.org/ocp/builder:rhel-8-golang-1.15-openshift-4.6`. How do they relate?

Can you link the github repo where these image streams are defined? (the one from Slack)

Reply:
So the openshift/release:golang-$VERSION images that seem to be often used as build_root are defined here: https://github.com/openshift/release/tree/master/projects/origin-release. Those are based on centos and install a plenthea of other stuff. The `ocp/builder:rhel-8-golang-1.15-openshift-4.6` images come from ART I believe.

tbh I have no idea why we use the one image for build_root and the other one for the promotion and that certainly looks like it may result in surprises. @jupierce do you know?

Reply:
As part of the ubi8 migration fun, ART began to build a rhel based `release` image to replace the old centos images. This historical naming, "release," is confusing, so I prefer to call it a `build_root` image. This build_root image is a  golang builder + tools that upstream has come to expect during their CI tests. These tools are only used for CI tests and many of them are not available behind the firewall.

To produce an analog for these images, ART now creates buildconfigs and istags dynamically on api.ci to layer the CI-only tools atop each new golang builder image. 
https://github.com/openshift/doozer/blob/master/doozerlib/cli/streams_transforms/rhel-8/ci-build-root/Dockerfile captures much of what is layered on top of the ART golang builder image to create a build_root image.

The resulting images are out on api.ci with the following naming convention:
`registry.svc.ci.openshift.org/openshift/release:rhel-{7,8}-release-golang-1.{GO_MINOR}-openshift-{OCP_MAJOR}.{OCP_MINOR}`

e.g. registry.svc.ci.openshift.org/openshift/release:rhel-8-release-golang-1.15-openshift-4.6

fwiw, if an upstream repo is using an ART golang builder image (e.g. `registry.svc.ci.openshift.org/ocp/builder:rhel-8-golang-1.15-openshift-4.6`), the name of the appropriate release image to use is a trivial string manipulation away.